How many late days are you using for this assignment? 0

In [82]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [83]:
# Load the numerical dataset
numericalDataset = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Load the categorical dataset
categoricalDataset = pd.read_csv('chronic_kidney_disease_categorical.csv')

# Print the datasets
print(f"Numerical Dataset:\n{numericalDataset.head()}")
print(f"Categorical Dataset:\n{categoricalDataset.head()}")

Numerical Dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  8200.0   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN     NaN   

   rbcc  
0   6.4  
1   5.5  
2   5.7  
3   4.5  
4   NaN  
Categorical Dataset:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0   

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data in this dataset are numerical and categorical health metrics that may pertain to detecting chronic kidney disease, including blood pressure, specific gravity, albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, red blood cell count, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, and anemia. These metrics would be called features if they are relevant to identifying, or labeling, a patient as having chronic kidney disease. Thus, the label is a boolean indicating whether or not the patient has CKD. 3 examples of categorical columns are sg (specific gravity), al (albumin), and su (sugar). 3 examples of numerical columns are age, bp (blood pressure), and bgr (blood glucose random).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [84]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numericalDuplicateRows = numericalDataset.duplicated(subset=['unique_id']).sum()
print(f"Duplicate rows in numerical dataset: {numericalDuplicateRows}")


# Drop these duplicate rows
numericalDataset = numericalDataset.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categoricalDuplicateRows = categoricalDataset.duplicated(subset=['unique_id']).sum()
print(f"Duplicate rows in categorical dataset: {categoricalDuplicateRows}")
categoricalDataset = categoricalDataset.drop_duplicates()


Duplicate rows in numerical dataset: 30
Duplicate rows in categorical dataset: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: Not all unique IDs are present in both datasets. This may be because this data wasn't collected from some patients. In the combined table, there are missing data points in some rows.

In [85]:
# Print number of unique ids in both datasets
print(f"Number of unique IDs in numerical dataset: {numericalDataset['unique_id'].nunique()}")
print(f"Number of unique IDs in categorical dataset: {categoricalDataset['unique_id'].nunique()}")
print(f"Number of unique IDs common to both datasets: {numericalDataset['unique_id'].isin(categoricalDataset['unique_id']).sum()}")

# Merge the two given numerical and categorical datasets based on their unique_ID.
mergedDataset = pd.merge(numericalDataset, categoricalDataset, on='unique_id')

#Print the combined dataset
print(f"Number of unique IDs in merged dataset: {mergedDataset['unique_id'].nunique()}")
print(f"Merged dataset:\n{mergedDataset.head()}")

Number of unique IDs in numerical dataset: 390
Number of unique IDs in categorical dataset: 400
Number of unique IDs common to both datasets: 390
Number of unique IDs in merged dataset: 390
Merged dataset:
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  ...   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
3     764115  70.0  90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
4     240975  47.0  80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN  ...   

       pc         pcc          ba  htn   dm cad appet   pe ane  Target  
0  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
1  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
2  normal  notpresent  notpresent   no   no  no  good   no  no  notckd  
3  normal  notpresent  notpresent  yes  yes  

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [86]:
# Calculate the percentage of rows that contain atleast one missing value
incompleteRowPercentage = mergedDataset.isnull().any(axis=1).sum() / len(mergedDataset) * 100

# Print %
print(f"Percentage of rows with incomplete data: {incompleteRowPercentage:.2f}%")

# Drop these rows from the dataset
noNullsDataset = mergedDataset.dropna()

# Print the Dataset
print(f"Dataset after dropping incomplete rows:\n{noNullsDataset.head()}")

Percentage of rows with incomplete data: 60.77%
Dataset after dropping incomplete rows:
    unique_id   age    bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0      203694  38.0  80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1      938027  43.0  60.0  108.0  25.0  1.0  144.0  5.0  17.8  43.0  ...   
2      421471  37.0  60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
7      109053  64.0  60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
13     349892  59.0  70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   

          pc         pcc          ba  htn   dm  cad appet  pe ane  Target  
0     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
1     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
2     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
7     normal  notpresent  notpresent   no   no   no  good  no  no  notckd  
13  abnormal  notpresent  notpresent  yes  yes  yes  good  no  no     ckd  

## Part 3.4: Sort the dataset according to the Labels

In [87]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sortedDataset = noNullsDataset.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
print(f"Sorted dataset:\n{sortedDataset.head()}")

Sorted dataset:
   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

         pc         pcc          ba  htn   dm  cad appet   pe  ane Target  
0  abnormal     present     present   no   no   no  good   no  yes    ckd  
1  abnormal     present  notpresent  yes  yes  yes  poor   no   no    ckd  
2  abnormal     present  notpresent  yes  yes   no  poor   no  yes    ckd  
3  abnormal  notpresent  notpresent  yes  yes  yes  poor  yes  yes    ckd  
4    normal  notpresent  notpresent   no   no   no  good   no   no    ckd  

[5 rows x 25 columns]


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [88]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encodedDataset = pd.get_dummies(sortedDataset, columns=categorical_columns, drop_first=True)

# Print the dataset
print(f"Encoded dataset:\n{encodedDataset.head()}")

Encoded dataset:
   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
2     484175  53.0   90.0   70.0  107.0  7.2  114.0  3.7   9.5  29.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   

   pc_normal  pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  \
0      False         True        True    False   False    False       False   
1      False         True       False     True    True     True        True   
2      False         True       False     True    True    False        True   
3      False        False       False     True    True     True        True   
4       True        False       False    False   False    False       False   

   pe_yes  ane_yes  Target_notckd  
0   False

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Conceptually, both solutions are equivalent. For the downstream modeling, however, this may introduce multicollinearity and biased results. For example, if a dataset had columns "diabetes_yes", "diabetes_no", and "age", then the redundant diabetes column would double the influence of this metric in downstream modeling. This applies to enumerated categories too, like "type of condiment". Thus, it would be wise to pass `drop_first=True` to `.get_dummies()`, which would eliminate redundant variables.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [89]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
noOutliersDataset = encodedDataset[(encodedDataset[numerical_columns].sub(encodedDataset[numerical_columns].mean()).div(encodedDataset[numerical_columns].std()).abs().le(3).all(axis=1))]
# Print the dataset
print(f"Removed {len(encodedDataset) - len(noOutliersDataset)} outliers")
print(f"Dataset after removing outliers:\n{noOutliersDataset.head()}")

Removed 19 outliers
Dataset after removing outliers:
   unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1     137148  73.0  100.0  295.0   90.0  5.6  140.0  2.9   9.2  30.0  ...   
3     343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4     532520  60.0   90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
6     546225  46.0   60.0  163.0   92.0  3.3  141.0  4.0   9.8  28.0  ...   

   pc_normal  pcc_present  ba_present  htn_yes  dm_yes  cad_yes  appet_poor  \
0      False         True        True    False   False    False       False   
1      False         True       False     True    True     True        True   
3      False        False       False     True    True     True        True   
4       True        False       False    False   False    False       False   
6       True        False       False     True    True    False       False   

   pe_yes

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [90]:
# Normalize the all Numerical Attributes in the dataset.
# Min-Max Normalization
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
normalizedDataset = encodedDataset.copy()
normalizedDataset[numerical_columns] = (encodedDataset[numerical_columns] - encodedDataset[numerical_columns].min()) / (encodedDataset[numerical_columns].max() - encodedDataset[numerical_columns].min())

# Print the dataset
print(f"Normalized dataset:\n{normalized_dataset.head()}")

Normalized dataset:
   unique_id       age        bp       bgr        bu        sc       sod  \
0     474407  0.194805  0.666667  0.088095  0.100334  0.087838  0.358974   
1     137148  0.870130  0.833333  0.535714  0.267559  0.351351  0.743590   
2     484175  0.610390  0.666667  0.000000  0.324415  0.459459  0.076923   
3     343710  0.714286  0.500000  0.245238  0.461538  0.236486  0.615385   
4     532520  0.701299  0.666667  0.083333  0.143813  0.128378  0.641026   

        pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
0  0.022472  0.353741  0.311111  ...      False         True        True   
1  0.008989  0.414966  0.466667  ...      False         True       False   
2  0.026966  0.435374  0.444444  ...      False         True       False   
3  0.060674  0.312925  0.333333  ...      False        False       False   
4  0.060674  0.544218  0.533333  ...       True        False       False   

   htn_yes  dm_yes  cad_yes  appet_poor  pe_yes  ane_yes  Target_n

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The `unique_id` column can be removed as it does not contain any information relevant to modeling chronic kidney disease from health metrics. It was useful for prepping the data, but it is not useful for modeling.

In [91]:
#Remove that column
finalDataset = normalizedDataset.drop(columns=['unique_id'])

# Print the dataset
print(f"Final dataset:\n{finalDataset.head()}")

Final dataset:
        age        bp       bgr        bu        sc       sod       pot  \
0  0.194805  0.666667  0.088095  0.100334  0.087838  0.358974  0.022472   
1  0.870130  0.833333  0.535714  0.267559  0.351351  0.743590  0.008989   
2  0.610390  0.666667  0.000000  0.324415  0.459459  0.076923  0.026966   
3  0.714286  0.500000  0.245238  0.461538  0.236486  0.615385  0.060674   
4  0.701299  0.666667  0.083333  0.143813  0.128378  0.641026  0.060674   

       hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \
0  0.353741  0.311111  0.380531  ...      False         True        True   
1  0.414966  0.466667  0.141593  ...      False         True       False   
2  0.435374  0.444444  0.367257  ...      False         True       False   
3  0.312925  0.333333  0.238938  ...      False        False       False   
4  0.544218  0.533333  0.296460  ...       True        False       False   

   htn_yes  dm_yes  cad_yes  appet_poor  pe_yes  ane_yes  Target_notckd  
0  

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [92]:
# Export the dataframe to a new csv file
finalDataset.to_csv('final_ckd.csv', index=False)
# Export the dataframe to a new json file
finalDataset.to_json('final_ckd.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [96]:
print(f"Top 15 rows of the final dataset:\n{finalDataset.head(15)}")

Top 15 rows of the final dataset:
         age        bp       bgr        bu        sc       sod       pot  \
0   0.194805  0.666667  0.088095  0.100334  0.087838  0.358974  0.022472   
1   0.870130  0.833333  0.535714  0.267559  0.351351  0.743590  0.008989   
2   0.610390  0.666667  0.000000  0.324415  0.459459  0.076923  0.026966   
3   0.714286  0.500000  0.245238  0.461538  0.236486  0.615385  0.060674   
4   0.701299  0.666667  0.083333  0.143813  0.128378  0.641026  0.060674   
5   0.558442  0.833333  0.140476  0.494983  0.770270  0.282051  0.015730   
6   0.519481  0.166667  0.221429  0.274247  0.195946  0.769231  0.033708   
7   0.636364  0.666667  0.483333  0.752508  0.932432  0.538462  0.020225   
8   0.649351  0.666667  0.140476  0.324415  0.425676  0.512821  0.051685   
9   1.000000  0.333333  0.076190  0.167224  0.148649  0.102564  0.071910   
10  0.688312  0.500000  0.554762  0.083612  0.060811  0.282051  0.022472   
11  0.870130  0.333333  0.000000  0.073579  0.033784  

Paste here: 

### Paste the markdown table here
|   age   |    bp    |    bgr   |    bu    |    sc    |    sod   |   pot    |   hemo   |    pcv   |    wbcc  | pc_normal | pcc_present | ba_present | htn_yes | dm_yes | cad_yes | appet_poor | pe_yes | ane_yes | Target_notckd |
|---------|----------|----------|----------|----------|----------|----------|----------|----------|----------|-----------|-------------|------------|---------|--------|---------|------------|--------|---------|----------------|
| 0.194805| 0.666667 | 0.088095 | 0.100334 | 0.087838 | 0.358974 | 0.022472 | 0.353741 | 0.311111 | 0.380531 | False     | True        | True       | False   | False  | False   | False      | False  | True    | False          |
| 0.870130| 0.833333 | 0.535714 | 0.267559 | 0.351351 | 0.743590 | 0.008989 | 0.414966 | 0.466667 | 0.141593 | False     | True        | False      | True    | True   | True    | True       | False  | False   | False          |
| 0.610390| 0.666667 | 0.000000 | 0.324415 | 0.459459 | 0.076923 | 0.026966 | 0.435374 | 0.444444 | 0.367257 | False     | True        | False      | True    | True   | False   | True       | False  | True    | False          |
| 0.714286| 0.500000 | 0.245238 | 0.461538 | 0.236486 | 0.615385 | 0.060674 | 0.312925 | 0.333333 | 0.238938 | False     | False       | False      | True    | True   | True    | True       | True   | True    | False          |
| 0.701299| 0.666667 | 0.083333 | 0.143813 | 0.128378 | 0.641026 | 0.060674 | 0.544218 | 0.533333 | 0.296460 | True      | False       | False      | False   | False  | False   | False      | False  | False   | False          |
| 0.558442| 0.833333 | 0.140476 | 0.494983 | 0.770270 | 0.282051 | 0.015730 | 0.340136 | 0.333333 | 0.256637 | False     | False       | False      | True    | True   | False   | True       | True   | True    | False          |
| 0.519481| 0.166667 | 0.221429 | 0.274247 | 0.195946 | 0.769231 | 0.033708 | 0.455782 | 0.422222 | 0.477876 | True      | False       | False      | True    | True   | False   | False      | False  | False   | False          |
| 0.636364| 0.666667 | 0.483333 | 0.752508 | 0.932432 | 0.538462 | 0.020225 | 0.353741 | 0.288889 | 0.477876 | False     | False       | False      | True    | True   | False   | True       | True   | True    | False          |
| 0.649351| 0.666667 | 0.140476 | 0.324415 | 0.425676 | 0.512821 | 0.051685 | 0.408163 | 0.444444 | 0.115044 | False     | False       | False      | True    | False  | False   | False      | False  | False   | False          |
| 1.000000| 0.333333 | 0.076190 | 0.167224 | 0.148649 | 0.102564 | 0.071910 | 0.380952 | 0.377778 | 0.398230 | True      | False       | False      | True    | False  | False   | True       | False  | True    | False          |
| 0.688312| 0.500000 | 0.554762 | 0.083612 | 0.060811 | 0.282051 | 0.022472 | 0.496599 | 0.577778 | 0.314159 | True      | False       | False      | False   | True   | False   | True       | False  | False   | False          |
| 0.870130| 0.333333 | 0.000000 | 0.073579 | 0.033784 | 0.358974 | 0.033708 | 0.469388 | 0.444444 | 0.668142 | True      | False       | False      | True    | True   | False   | False      | True   | False   | False          |
| 0.688312| 0.000000 | 0.407143 | 0.605351 | 0.783784 | 0.076923 | 0.008989 | 0.442177 | 0.488889 | 0.526549 | False     | False       | False      | False   | True   | False   | False      | True   | False   | False          |
| 0.701299| 0.166667 | 0.519048 | 0.086957 | 0.087838 | 0.487179 | 0.011236 | 0.326531 | 0.355556 | 0.504425 | False     | True        | False      | True    | False  | False   | True       | False  | True    | False          |
| 0.701299| 0.666667 | 1.000000 | 0.284281 | 0.155405 | 0.512821 | 0.029213 | 0.571429 | 0.577778 | 0.362832 | True      | False       | True       | True    | True   | False   | False      | False  | False   | False          |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer - To mitigate data conversion hallucinations, you could use an LLM to generate a converter script that does the format transformation instead of relying on the LLM for the transformation itself. Or you could ask it to generate a script that verifies data integrity after the transformation by an LLM. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [97]:
# Code to rename all the columns in the dataset

# Define the mappings for renaming columns
column_mappings = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns
finalDataset.rename(columns=column_mappings, inplace=True)

# Display the updated DataFrame (optional)
print(finalDataset.head())

        age  blood pressure  blood glucose random  blood urea  \
0  0.194805        0.666667              0.088095    0.100334   
1  0.870130        0.833333              0.535714    0.267559   
2  0.610390        0.666667              0.000000    0.324415   
3  0.714286        0.500000              0.245238    0.461538   
4  0.701299        0.666667              0.083333    0.143813   

   serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0          0.087838  0.358974   0.022472    0.353741            0.311111   
1          0.351351  0.743590   0.008989    0.414966            0.466667   
2          0.459459  0.076923   0.026966    0.435374            0.444444   
3          0.236486  0.615385   0.060674    0.312925            0.333333   
4          0.128378  0.641026   0.060674    0.544218            0.533333   

       wbcc  ...  pc_normal  pcc_present  ba_present  htn_yes  dm_yes  \
0  0.380531  ...      False         True        True    False   False   
1  0.1

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [99]:
# Converted SQL to Pandas code

# Group by 'Target' and count occurrences
result = finalDataset.groupby('Target_notckd').size().reset_index(name='count')

# Display the resulting DataFrame
print(result)


   Target_notckd  count
0          False     42
1           True    111


*Describe what the above code does here*

Answer - The above code groups the dataset according to the `Target_notckd` column and counts the number of rows each group has, then prints the output.